lib/db_utils.go

```diff
@@ -1204,26 +1215,39 @@ func DBDeleteWithTxn(txn *badger.Txn, snap *Snapshot, key []byte, eventManager *
 	var getError error
 	isState := snap != nil && snap.isState(key)
 
+	isCoreState := isCoreStateKey(key)
+
 	// If snapshot was provided, we will need to load the current value of the record
 	// so that we can later write it in the ancestral record. We first lookup cache.
-	if isState {
-		// We check if we've already read this key and stored it in the cache.
-		// Otherwise, we fetch the current value of this record from the DB.
-		ancestralValue, getError = DBGetWithTxn(txn, snap, key)
-		// If the key doesn't exist then there is no point in deleting this entry.
-		if getError == badger.ErrKeyNotFound {
-			return nil
+	if isState || (isCoreState && eventManager != nil && eventManager.isMempoolManager) {
+		// When we are syncing state from the mempool, we need to read the last committed view txn.
+		// This is because we will be querying the badger DB, and during the flush loop, every entry that is
+		// updated will first be deleted. In order to counteract this, we reference a badger transaction that was
+		// initiated before the flush loop started.
+		if eventManager != nil && eventManager.isMempoolManager && eventManager.lastCommittedViewTxn != nil {
+			ancestralValue, getError = DBGetWithTxn(eventManager.lastCommittedViewTxn, snap, key)
+		} else {
+			// We check if we've already read this key and stored it in the cache.
+			// Otherwise, we fetch the current value of this record from the DB.
+			ancestralValue, getError = DBGetWithTxn(txn, snap, key)
+			// If the key doesn't exist then there is no point in deleting this entry.
+			if getError == badger.ErrKeyNotFound {
+				return nil
+			}
 		}
 
 		// If there is some error with the DB read, other than non-existent key, we return.
-		if getError != nil {
+		if getError != nil && getError != badger.ErrKeyNotFound {
 			return errors.Wrapf(getError, "DBDeleteWithTxn: problem checking for DB record "+
 				"with key: %v", key)
 		}
 	}
 
 	err := txn.Delete(key)
-	if err != nil {
+	if err != nil && err == badger.ErrKeyNotFound && eventManager != nil && eventManager.isMempoolManager {
+		// If the key doesn't exist then there is no point in deleting this entry.
+		return nil
```

Comment:
Was the previous code causing issues for you? just curious. I don't see any issue w/ this change, but it does change the intent a little bit.

Reply:
This is specifically logic just for the mempool manager, which is only active during the mempool state sync routine. I am struggling to remember the exact circumstances here, but in some instances when deleting in the state syncer mempool routine, we need these extra checks here.

Reply:
works for me!